File: src/main/java/io/transwarp/logparser/filter/TimeFilter.java
package io.transwarp.logparser.filter;

import io.transwarp.logparser.util.Record;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * Author: stk
 * Date: 2018/4/3
 */
public class TimeFilter implements Filter {
    private Date beginTime;
    private Date endTime;

    public TimeFilter(String beginTime, String endTime) {
        SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd HH:mm:ss,SSS");
        try {
            this.beginTime = format.parse(beginTime);
            this.endTime = format.parse(endTime);
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean filter(Record record) {
        Date current = record.getTime();
        if (current == null) return false;
        return current.compareTo(beginTime) >= 0 && current.compareTo(endTime) <= 0;
    }
}

File: src/main/scala/io/transwarp/logparser/filter/TimeFilter.scala
package io.transwarp.logparser.filter

import java.text.SimpleDateFormat
import java.util.Date

import io.transwarp.logparser.util.LogEntry

/**
  * Author: stk
  * Date: 2018/4/4
  *
  * @param params [begin time, end time]
  */
class TimeFilter(override val params: List[String]) extends Filter {
  private val format = new SimpleDateFormat("yyyyMMdd HH:mm:ss,SSS")

  override def filter(logEntry: LogEntry): Boolean = logEntry.format("timestamp").asInstanceOf[Option[Date]]
    .fold(false)(f => f.compareTo(format.parse(params.head)) >= 0 && f.compareTo(format.parse(params.last)) <= 0)

  override def copy(): Filter = new TimeFilter(params)
}
